src/wire/ipv6.rs

```diff
@@ -0,0 +1,413 @@
+    pub fn from_bytes(data: &[u8]) -> Address {
+        let mut bytes = [0; 16];
+        bytes.copy_from_slice(data);
+        Address(bytes)
+    }
+
+    /// Construct an IPv6 address from a sequence of words, in big-endian.
+    ///
+    /// # Panics
+    /// The function panics if `data` is not 8 words long.
+    pub fn from_parts(data: &[u16]) -> Address {
+        assert!(data.len() >= 8);
+        let mut bytes = [0; 16];
+        for word_idx in 0..8 {
+            let byte_idx = word_idx * 2;
+            NetworkEndian::write_u16(&mut bytes[byte_idx..(byte_idx + 2)], data[word_idx]);
+        }
+        Address(bytes)
+    }
+
+    /// Write a IPv6 address to the given slice.
+    ///
+    /// # Panics
+    /// The function panics if `data` is not 8 words long.
+    pub fn write_parts(&self, data: &mut [u16]) {
+        assert!(data.len() >= 8);
+        for i in 0..8 {
+            let byte_idx = i * 2;
+            data[i] = NetworkEndian::read_u16(&self.0[byte_idx..(byte_idx + 2)]);
+        }
+    }
+
+    /// Return an IPv6 address as a sequence of octets, in big-endian.
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.0
+    }
+
+    /// Query whether the IPv6 address is an unicast address.
+    pub fn is_unicast(&self) -> bool {
+        !(self.is_multicast() || self.is_unspecified())
+    }
+
+    /// Query whether the IPv6 address is a multicast address.
+    pub fn is_multicast(&self) -> bool {
+        self.0[0] == 0xff
+    }
+
+    /// Query whether the IPv6 address is the "unspecified" address.
+    pub fn is_unspecified(&self) -> bool {
+        self.0 == [0x00; 16]
+    }
+
+    /// Query whether the IPv6 address is in the "link-local" range.
+    pub fn is_link_local(&self) -> bool {
+        self.0[0..8] == [0xfe, 0x80, 0x00, 0x00,
+                         0x00, 0x00, 0x00, 0x00]
+    }
+
+    /// Query whether the IPv6 address is the "loopback" address.
+    pub fn is_loopback(&self) -> bool {
+        *self == Self::LOOPBACK
+    }
+
+    /// Helper function used to mask an addres given a prefix.
+    ///
+    /// # Panics
+    /// This function panics if `mask` is greater than 128.
+    pub(super) fn mask(&self, mask: u8) -> [u8; 16] {
+        assert!(mask <= 128);
+        let mut bytes = [0u8; 16];
+        let idx = (mask as usize) / 8;
+        let modulus = (mask as usize) % 8;
+        let (first, second) = self.0.split_at(idx);
+        bytes[0..idx].copy_from_slice(&first);
+        if idx < 16 {
+            let part = second[0];
+            bytes[idx] = part & (!(0xff >> modulus) as u8);
+        }
+        bytes
+    }
+}
+
+impl fmt::Display for Address {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        if self.0[0..10] == [0u8; 10] && self.0[10..12] == [0xffu8; 2] {
+            // The address is a IPv4 mapped address
+            write!(f, "::ffff:{}.{}.{}.{}", self.0[12], self.0[13], self.0[14], self.0[15])
```

Comment:
Seems weird to stringify those since we cannot parse them. It should be one or the other.

Reply:
After thinking about this again, adding IPv4 mapped addresses could be a good follow up PR. AFAIK `From<Ipv4Address>` should be implemented for `Ipv6Addr`, we should be able to parse IPv4 mapped addresses, and we should print them differently. I'll remove this for now